### ISSUE TITLE
[css-text4] text-spacing-trim and classes of 

### ISSUE THREAD

**Issue by frivoal** [w3c/csswg-drafts#9504]
In [css-text-4 8.5.1. Fullwidth Punctuation Collapsing](https://drafts.csswg.org/css-text-4/#fullwidth-collapsing), we have:

> Set fullwidth closing punctuation half-width if the next character is any of:
> * a fullwidth closing punctuation
> * a fullwidth middle dot punctuation
> * an ideographic space (U+3000)
> * a fullwidth opening punctuation of a larger font-size
> * a character belonging to Unicode general category Pe

I think we should we add `Pf` alongside with `Pe`.

These WPT tests assume it, and that seems very reasonable to me:
https://wpt.fyi/results/css/css-text/text-spacing-trim?label=master&label=experimental&aligned&q=text-spacing-trim-quote-001.html

Further, thoes tests makes it conditional on the language, collapsing when it's Japanese, but not when it's Chinese. If we were to do spec that language dependency, it should be for more than just `Pf`, but also `Pe` and `Ps` (in the previous rule). Or maybe we don't need the language dependency, and the variants of those glyphs that can or cannot be trimmed can be detected from the font?

cc: @fantasai @kojiishi

**Comment by kojiishi** [comment]
> These WPT tests assume it

The "[fullwidth opening punctuation](https://drafts.csswg.org/css-text-4/#fullwidth-opening-punctuation)" includes quotation marks, so the test matches the spec without needing to including `Pf`.

Aside that, wether to include [`Pf`](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%3Agc%3DPf%3A%5D&g=&i=) or not...I'm not sure. They are not commonly used in CJK context, and not all of them look like open nor close to me, so I guess it's not necessary, but not strongly oppose.

> Further, thoes tests makes it conditional on the language, collapsing when it's Japanese, but not when it's Chinese. If we were to do spec that language dependency,...

The test uses a web font (Noto CJK) and it's written specifically for that font, without considering other fonts. Noto CJK follows a variation of the convention often used by Adobe fonts (see the [official font readme file](https://github.com/adobe-fonts/source-han-sans/raw/release/SourceHanSansReadMe.pdf) for more details.) But there are surprisingly many other variants that I'm hesitant to spec all those variants.

**Comment by frivoal** [comment]
> The "[fullwidth opening punctuation](https://drafts.csswg.org/css-text-4/#fullwidth-opening-punctuation)" includes quotation marks, so the test matches the spec without needing to including `Pf`.

You're right, sorry, I had missed that part. That makes my first point irrelevant.


> > Further, thoes tests makes it conditional on the language, collapsing when it's Japanese, but not when it's Chinese. If we were to do spec that language dependency,...
> 
> The test uses a web font (Noto CJK) and it's written specifically for that font, without considering other fonts. Noto CJK follows a variation of the convention often used by Adobe fonts (see the [official font readme file](https://github.com/adobe-fonts/source-han-sans/raw/release/SourceHanSansReadMe.pdf) for more details.) But there are surprisingly many other variants that I'm hesitant to spec all those variants.

Does that mean that Chrome is effectively depending on the spec's allowance for not trimming in the case of proportional fonts, by considering these characters proportional in the languages where they don't render as full-width? Or something else?

**Comment by kojiishi** [comment]
Hmm...I'm not sure if I understand all accurate nuance of your question, sorry for my English skill, but IIUC, the answer is yes. The spec requires applying only to fullwidth glyphs, right? The current implementation in Chromium applies the kerning by applying `halt`. The `halt` should have no effects for propotional glyphs by definition.

The current implementation also checks if glyphs are fullwidth or not for some cases, and also ink is at the expected part of the glyph spaces. This is needed for some other cases and for the performance optimizations, but this part will need further tuning by testing more fonts. The current implementation adds about ~7% cost to the shape engine, I'm trying to reduce this a bit more.

Does this answer all your questions?

**Comment by kojiishi** [comment]
Note that `Pi` and `Pf` are:
> [Characters with the quotation-related General_Category values Pi or Pf may behave like opening punctuation (gc=Ps) or closing punctuation (gc=Pe), depending on usage and quotation conventions.](https://unicode.org/reports/tr44/#:~:text=Characters%20with%20the%20quotation%2Drelated%20General_Category%20values%20Pi%20or%20Pf%20may%20behave%20like%20opening%20punctuation%20(gc%3DPs)%20or%20closing%20punctuation%20(gc%3DPe)%2C%20depending%20on%20usage%20and%20quotation%20conventions.)

So if there are code points that should be included, we may want to include selectively.

**Comment by frivoal** [comment]
> The current implementation also checks if glyphs are fullwidth or not for some cases,

This is spec compliant.

> and also ink is at the expected part of the glyph spaces.

I'm not sure the spec covers that case, but maybe it should? Based on the way I read it, you are allowed (but not required) to use ink detection instead of `halt` and `vhal` to find where is the white space you need to trim, but I don't think the spec allows you to use ink detection to decide if you want to trim at all.

**Comment by kojiishi** [comment]
Well, re-reading the spec, I think you're right. Avoiding glyph collisions is the top priority for me, so I guess I misread the spec.

There are two places in the spec. The [8.5.2. Text Spacing Character Classes](https://drafts.csswg.org/css-text-4/#fullwidth-collapsing) requires it only for colon and dots, but not for other types.

In antother place, [8.5.1. Fullwidth Punctuation Collapsing](https://drafts.csswg.org/css-text-4/#fullwidth-collapsing) says:
> trim (kern) the blank half of the glyphs, if they are given full-width and must be set half-width

This may read impls need to check inks as it says it can trim only "blank half", but it can improve, such as what to do if the glyph doesn't have "blank half".

For the latter, it says "must be set half-width". We apply `halt`, which doesn't guarantee to set half-width. This happens with [U+FF5D FULLWIDTH RIGHT CURLY BRACKET](https://codepoints.net/U+ff5d) in extra-bold fonts when ink fills a bit more than half, and thus font designers set its `halt` a bit wider than half-width.

/cc @nt1m @vitorroriz @clqsin45
